### PR TITLE
Fallback to serial execution if parallel execution fails

### PR DESCRIPTION
**Pull request by cffls**
# Description

Enables the node to bypass unexpected errors from parallel execution and use the result from serial execution instead.

# Changes

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Changes only for a subset of nodes

# Nodes audience

In case this PR includes changes that must be applied only to a subset of nodes, please specify how you handled it (e.g. by adding a flag with a default value...)

# Checklist

- [ ] I have added at least 2 reviewer or the whole pos-v1 team
- [ ] I have added sufficient documentation in code
- [ ] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply
- [ ] Created a task in Jira and informed the team for implementation in Erigon client (if applicable)
- [ ] Includes RPC methods changes, and the Notion documentation has been updated

# Cross repository changes

- [ ] This PR requires changes to heimdall
    - In case link the PR here:
- [ ] This PR requires changes to matic-cli
    - In case link the PR here:

## Testing

- [x] I have added unit tests
- [ ] I have added tests to CI
- [x] I have tested this code manually on local environment
- [ ] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on mumbai/amoy
- [ ] I have created new e2e tests into express-cli

### Manual tests

Please complete this section with the steps you performed if you ran manual tests for this functionality, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -606,44 +609,58 @@ func (bc *BlockChain) ProcessBlock(block *types.Block, parent *types.Header) (_
...
+				vtime = time.Since(vstart)
+			}
+			resultChan <- Result{receipts, logs, usedGas, err, statedb, blockExecutionSerialCounter, false}
 		}()
 	}
 
 	result := <-resultChan
 
-	if _, ok := result.err.(blockstm.ParallelExecFailedError); ok {
+	if result.parallel && result.err != nil {
```
Comment (Raneet10) [inline core/blockchain.go:657]: We're almost always guaranteed to have the parallel execution result first in `resultChan` right ? By common sense it should be that way but just ensuring.

File: core/blockchain.go [hunk 2]
```diff
@@ -606,44 +609,58 @@ func (bc *BlockChain) ProcessBlock(block *types.Block, parent *types.Header) (_
...
 		log.Warn("Parallel state processor failed", "err", result.err)
-
+		blockExecutionParallelErrorCounter.Inc(1)
 		// If the parallel processor failed, we will fallback to the serial processor if enabled
 		if processorCount == 2 {
-			result.statedb.StopPrefetcher()
 			result = <-resultChan
+			result.statedb.StopPrefetcher()
```
Comment (Raneet10) [inline core/blockchain.go:663]: So if parallel execution doesn't fail we don't stop the prefetcher ? There's a clause below that seems to ensure we shut down both serial and parallel's prefetchers but my assumption could be wrong.

**Review by Raneet10 — APPROVED** [review]
LGTM! Just some minor doubts.

File: core/blockchain.go (hunk 1 above)
Comment (cffls) [reply core/blockchain.go:657]: Yes, we will mostly get parallel result. However, in the case where the result is returned from serial, this check `if result.parallel` will skip the code block.

File: core/blockchain.go (hunk 2 above)
Comment (cffls) [reply core/blockchain.go:663]: Yes, the one below will shutdown the remaining prefetcher, regardless what's left.

Comment (Raneet10) [reply core/blockchain.go:663]: I don't think I understood perfectly. If the parallel execution succeeds, we only end up stopping the serial's prefetcher. Is that expected ?

Comment (cffls) [reply core/blockchain.go:663]: Ahh let me try to clarify.. We will stop the prefetcher of the first finished executor at line 663 (the one you highlighted), and then stop the second prefetcher at [line 674](https://github.com/maticnetwork/bor/pull/1392/files#diff-53d5f4b8a536ec2a8c8c92bf70b8268f1d77ad77e9f316e6f68a2bcae5303215R674).

**Review by ssandeep — APPROVED** [review]

**Pull request merged**